dbms/src/Flash/Mpp/ReceiverChannelTryWriter.cpp

```diff
@@ -0,0 +1,152 @@
+
+    // Still need to send error_ptr or resp_ptr even if packet.chunks_size() is zero.
+    for (size_t i = 0; i < channel_size && loopJudge(res); ++i)
+    {
+        if (resp_ptr == nullptr && error_ptr == nullptr && chunks[i].empty())
+            continue;
+
+        std::shared_ptr<ReceivedMessage> recv_msg = std::make_shared<ReceivedMessage>(
+            source_index,
+            req_info,
+            tracked_packet,
+            error_ptr,
+            resp_ptr,
+            std::move(chunks[i]));
+
+        GRPCReceiveQueueRes write_res = tryWriteImpl(i, std::move(recv_msg));
+        if (write_res != GRPCReceiveQueueRes::OK)
+            res = write_res;
+
+        fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, res = GRPCReceiveQueueRes::CANCELLED);
+
+        // Only the first ExchangeReceiverInputStream need to handle resp.
+        resp_ptr = nullptr;
+    }
+    return res;
+}
+
+GRPCReceiveQueueRes ReceiverChannelTryWriter::tryWriteNonFineGrain(size_t source_index, const TrackedMppDataPacketPtr & tracked_packet, const mpp::Error * error_ptr, const String * resp_ptr)
+{
+    GRPCReceiveQueueRes res = GRPCReceiveQueueRes::OK;
+    auto & packet = tracked_packet->packet;
+    std::vector<const String *> chunks(packet.chunks_size());
+
+    for (int i = 0; i < packet.chunks_size(); ++i)
+        chunks[i] = &packet.chunks(i);
+
+    if (!(resp_ptr == nullptr && error_ptr == nullptr && chunks.empty()))
+    {
+        std::shared_ptr<ReceivedMessage> recv_msg = std::make_shared<ReceivedMessage>(
+            source_index,
+            req_info,
+            tracked_packet,
+            error_ptr,
+            resp_ptr,
+            std::move(chunks));
+
+        res = tryWriteImpl(0, std::move(recv_msg));
+        fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, res = GRPCReceiveQueueRes::CANCELLED);
+    }
+    return res;
+}
+
+template <bool enable_fine_grained_shuffle>
+GRPCReceiveQueueRes ReceiverChannelTryWriter::tryReWrite()
+{
+    GRPCReceiveQueueRes res = GRPCReceiveQueueRes::OK;
+    auto iter = rewrite_msgs.begin();
+
+    while (loopJudge(res) && (iter != rewrite_msgs.end()))
+    {
+        GRPCReceiveQueueRes write_res = tryRewriteImpl(iter->first, iter->second);
+        if (write_res == GRPCReceiveQueueRes::OK)
+        {
+            auto tmp_iter = iter;
+            ++iter;
+            rewrite_msgs.erase(tmp_iter);
+        }
+        else
+        {
+            ++iter;
+            res = write_res;
+        }
+
+        fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, res = GRPCReceiveQueueRes::CANCELLED);
+    }
+
+    return res;
+}
+
+GRPCReceiveQueueRes ReceiverChannelTryWriter::tryWriteImpl(size_t index, std::shared_ptr<ReceivedMessage> && msg)
+{
+    GRPCReceiveQueueRes res = grpc_recv_queues[index].push(msg);
+    if (res == GRPCReceiveQueueRes::FULL)
+        rewrite_msgs.insert({index, std::move(msg)});
```

Comment:
What will happen when the index is repeated? Or will it not happen? How about adding an assert?

Reply:
> What will happen when the index is repeated? Or will it not happen? How about adding an assert?

Same index is impossible to appear here, as only all write fail data has been written successfully the new data will goes on.
okk, I will add an assert